mongo/src/test/java/com/redhat/lightblue/mongo/crud/MongoCRUDControllerMemoryLimitsTest.java

```diff
@@ -0,0 +1,340 @@
+                query("{'field':'field1','op':'$eq','rvalue':'f1'}"),
+                update("{ '$set': { 'field2' : 'f2-updated' } }"),
+                projection("{'field':'*'}"));
+
+        Assert.assertEquals(10, response.getNumMatched());
+        Assert.assertEquals(10, response.getNumUpdated());
+
+        DBCursor cursor = db.getCollection("data").find();
+        assertTrue(cursor.hasNext());
+        cursor.forEach(obj -> {
+            String field2value = ((DBObject) obj).get("field2").toString();
+            assertEquals("f2-updated", field2value);
+        });
+        List<DocCtx> documents=streamToList(ctx);
+        Assert.assertEquals(10, documents.size());
+
+        Assert.assertTrue(ctx.getErrors().isEmpty());
+    }
+
+    /**
+     * Limit is above the needs of this update.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void updateDocuments_LimitNotExceeded() throws Exception {
+        setupTestDataAndMetadata();
+        insertDocs(10);
+
+        TestCRUDOperationContext ctx = new TestCRUDOperationContext("test", CRUDOperation.UPDATE);
+        ctx.add(emd);
+
+        ctx.getFactory().setMaxResultSetSizeForWritesB(11*testDocSize);
+        CRUDUpdateResponse response = controller.update(ctx,
+                query("{'field':'field1','op':'$eq','rvalue':'f1'}"),
+                update("{ '$set': { 'field2' : 'f2-updated' } }"),
+                projection("{'field':'*'}"));
+
+        Assert.assertEquals(10, response.getNumMatched());
+        Assert.assertEquals(10, response.getNumUpdated());
+
+        DBCursor cursor = db.getCollection("data").find();
+        assertTrue(cursor.hasNext());
+        cursor.forEach(obj -> {
+            String field2value = ((DBObject) obj).get("field2").toString();
+            assertEquals("f2-updated", field2value);
+        });
+        List<DocCtx> documents=streamToList(ctx);
+        Assert.assertEquals(10, documents.size());
+
+        Assert.assertTrue(ctx.getErrors().isEmpty());
+    }
+
+    /**
+     * Update accumulates a list of docs which exceeds the threshold.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void updateDocuments_LimitExceeded() throws Exception {
+
+        int count = MongoCRUDController.DEFAULT_BATCH_SIZE+10;
+
+        setupTestDataAndMetadata();
+        insertDocs(count);
+
+        TestCRUDOperationContext ctx = new TestCRUDOperationContext("test", CRUDOperation.UPDATE);
+        ctx.add(emd);
+
+        ctx.getFactory().setMaxResultSetSizeForWritesB((MongoCRUDController.DEFAULT_BATCH_SIZE+1)*testDocSize);
+        CRUDUpdateResponse response = controller.update(ctx,
+                query("{'field':'field1','op':'$eq','rvalue':'f1'}"),
+                update("{ '$set': { 'field2' : 'f2-updated' } }"),
+                projection("{'field':'*'}"));
+
+        // this is wrong - one batch was updated successfully
+        // see IterateAndUpdate.java:205 for more info
```

Comment:
This is ugly, but I think given our options it makes sense: "updated" probably should mean that hooks fired, too; that is the whole transaction was successful for those N documents. In this case, that's not true, so probably makes sense to consider them not "updated" despite the visible state change.